### PR TITLE
ramips: add support for D-Link DIR-878 A1

### PR DESCRIPTION
**Pull request by Sixtiz**
Specifications:
* SoC: MT7621AT
* RAM: 128MB
* Flash: 16MB NOR SPI flash
* WiFi: MT7615N (2.4GHz) and MT7615N (5Ghz)
* LAN: 5x1000M
* Firmware layout is Uboot with extra 96 bytes in header
* Base PCB is AP-MTKH7-0002
* LEDs Power Green,Power Orange,Internet Green,Internet Orange
* LEDs "2.4G" Green & "5G" Green connected directly to wifi module
* Buttons Reset,WPS,WIFI

Flashing instructions:

Upload image via emergency recovery mode
Push and hold reset button (on the back of the device) until power led starts flashing (about 10 secs or so) while powering the device on.
Give it ~30 seconds, to boot the recovery mode GUI
Connect your client computer to LAN1 of the device
Set your client IP address manually to 192.168.0.2 / 255.255.255.0.
Call the recovery page for the device at http://192.168.0.1
Use the provided emergency web GUI to upload and flash a new firmware to the device. Some browsers/OS combinations are known not to work, so if you don't see the percentage complete displayed and moving within a few seconds, restart the procedure from scratch and try anoher one, ot try the command line way.

Alternative method using command line on Linux:
curl -v -i -F "firmware=@openwrt-ramips-mt7621-dlink_dir-878-a1-squashfs-factory.bin" 192.168.0.1

Credit to @Lucky1openwrt for the patch. Original pull request: https://github.com/openwrt/openwrt/pull/2588/

Signed-off-by: Mathieu Martin-Borret <mathieu.mb@protonmail.com>

### CODE REVIEW

**Comment by Borromini** [comment]
Thanks @Sixtiz and @Lucky1openwrt.

Works fine here.

**Comment by adrianschmutzler** [comment]
The commit message lacks flashing instructions and the other information provided in the old PR ...

**Comment by Sixtiz** [comment]
> The commit message lacks flashing instructions and the other information provided in the old PR ...

Ok, I can add it back but that's just the standard instructions for D-Link routers as per the [wiki](https://openwrt.org/docs/guide-user/installation/installation_methods/d-link_recovery_gui), so is that really needed?

**Comment by Borromini** [comment]
@Sixtiz That is standard (and official) procedure yes. If the wiki gets adapted, wiped, lost, eaten by kittens - you can still find the instructions in the commit that adds support.

**Comment by Sixtiz** [comment]
@adrianschmutzler commit message updated.

**Comment by Borromini** [comment]
@Sixtiz mt7621.mk got modified so you'll need to rebase. The Build/ubootpad96 can be moved to the end of the Defines, I think (the order being alphabetical):
```

--- a/target/linux/ramips/image/mt7621.mk
+++ b/target/linux/ramips/image/mt7621.mk
@@ -105,6 +105,11 @@ define Build/ubnt-erx-factory-image
 	fi
 endef
 
+define Build/ubootpad96
+	uimage_padhdr -i $@ -o $@.new -l 96
+	mv $@.new $@
+endef
+
 define Device/afoundry_ew1200
   IMAGE_SIZE := 16064k
   DEVICE_VENDOR := AFOUNDRY
```

**Comment by Sixtiz** [comment]
> @Sixtiz mt7621.mk got modified so you'll need to rebase. The Build/ubootpad96 can be moved to the end of the Builld definitions, I think (the order being alphabetical):
Rebased and updated mt7621.mk

**Comment by Lucky1openwrt** [comment]
a change that you may need to know about
-  DEVICE_PACKAGES := kmod-mt7615e wpad-basic
+  DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware wpad-basic

https://git.openwrt.org/?p=openwrt/openwrt.git;a=blobdiff;f=target/linux/ramips/image/mt7621.mk;h=c2e96160cf7565eadc2b4a31c5242ebf82879468;hp=524364d8b0b3bfd6d60509325dd7f14e8d19999b;hb=81b59efefd4c88187a519d09129c97e63900731e;hpb=2dd26fda16a4574c91d02f4b1ccb2ce121e764c8

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts [hunk 1]
```diff
@@ -0,0 +1,180 @@
+/dts-v1/;
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:1]: Please add an SPDX license identifier.

Comment (Sixtiz) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts]: License added!

**Review by gauchim — APPROVED** [review]

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c [hunk 2]
```diff
@@ -407,6 +407,43 @@ static struct mtd_part_parser uimage_fonfxc_parser = {
 	.parse_fn = mtdsplit_uimage_parse_fonfxc,
 };
 
+/**************************************************
+ * ubootpad96
+ **************************************************/
+
+#define ubootpad96_PAD_LEN		96
```
Comment (ynezz) [inline target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:414]: Create separate commit for this changes with proper description. @rmilecki can you check this?

**Review by ynezz — CHANGES_REQUESTED** [review]
Create separate commit for the mtdsplit_uimage.c change, otherwise looks good to me, will wait for @rmilecki 's review and then merge.

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c (hunk 2 above)
Comment (Sixtiz) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:414]: Do you mean a separate commit in the same branch/pull request, or a new pull request?

Comment (ynezz) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:414]: Yes, this is good, thanks

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c [hunk 3]
```diff
@@ -407,6 +407,43 @@ static struct mtd_part_parser uimage_fonfxc_parser = {
...
+mtdsplit_uimage_parse_ubootpad96(struct mtd_info *master,
+			      const struct mtd_partition **pparts,
+			      struct mtd_part_parser_data *data)
+{
+	return __mtdsplit_parse_uimage(master, pparts, data,
+				       uimage_find_ubootpad96);
+}
+
+static const struct of_device_id mtdsplit_uimage_ubootpad96_of_match_table[] = {
+	{ .compatible = "ubootpad96,uimage" },
```
Comment (ynezz) [inline target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:436]: I would follow the naming of other parsers, so I would prefer `dlink,uimage`.

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c [hunk 4]
```diff
@@ -407,6 +407,43 @@ static struct mtd_part_parser uimage_fonfxc_parser = {
...
+	{},
+};
+
+static struct mtd_part_parser uimage_ubootpad96_parser = {
+	.owner = THIS_MODULE,
+	.name = "ubootpad96-fw",
```
Comment (ynezz) [inline target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:442]: `dlink-fw`

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c [hunk 5]
```diff
@@ -407,6 +407,43 @@ static struct mtd_part_parser uimage_fonfxc_parser = {
...
+	.of_match_table = mtdsplit_uimage_ubootpad96_of_match_table,
```
Comment (ynezz) [inline target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:443]: `mtdsplit_uimage_dlink_of_match_table`  and so on

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c [hunk 6]
```diff
@@ -407,6 +407,43 @@ static struct mtd_part_parser uimage_fonfxc_parser = {
 	.parse_fn = mtdsplit_uimage_parse_fonfxc,
 };
 
+/**************************************************
+ * ubootpad96
```
Comment (ynezz) [inline target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:411]: D-Link

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x20000>;
+				read-only;
+			};
+
+			partition@60000 {
+				compatible = "ubootpad96,uimage";
```
Comment (ynezz) [inline target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:104]: `compatible = "dlink,uimage";`

File: target/linux/ramips/image/mt7621.mk [hunk 8]
```diff
@@ -110,6 +110,11 @@ define Build/ubnt-erx-factory-image
 	fi
 endef
 
+define Build/ubootpad96
```
Comment (ynezz) [inline target/linux/ramips/image/mt7621.mk:113]: `dlink-uimage` or such

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c (hunk 3 above)
Comment (Lucky1openwrt) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:436]: not just dlink used in Motorola MR1700 & Motorola MR2600 
maybe use manufacturing company SGE (T&W) Shenzhen Gongjin Electronics 
https://deviwiki.com/wiki/D-Link_DIR-878_rev_A1

Comment (ynezz) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:436]: Ok, so `sge,uimage` and please mention the reason for the parser name in the commit message

**Comment by adrianschmutzler** [comment]
Commit message (2nd commit) still needs proper wrapping of lines (max. 74 chars per line).

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+			status = "okay";
+			label = "wan";
+			mtd-mac-address = <&factory 0xe006>;
+		};
+	};
+};
+
+&state_default {
+	gpio {
+		groups = "uart3", "jtag", "wdt";
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:178]: missing i2c for gpios 3 and 4

File: target/linux/ramips/image/mt7621.mk (hunk 8 above)
Comment (adrianschmutzler) [reply target/linux/ramips/image/mt7621.mk:113]: Please also take care of proper sorting when you update.

File: target/linux/ramips/image/mt7621.mk [hunk 10]
```diff
@@ -241,6 +246,21 @@ define Device/dlink_dir-860l-b1
...
+define Device/dlink_dir-878-a1
+  IMAGE_SIZE := 16000k
+  DEVICE_VENDOR := D-Link
+  DEVICE_MODEL := DIR-878
+  DEVICE_VARIANT := A1
+  DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware wpad-basic
+  KERNEL_INITRAMFS := $$(KERNEL) | ubootpad96
+  IMAGES += factory.bin
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | ubootpad96 |\
+	pad-rootfs | check-size | append-metadata
```
Comment (adrianschmutzler) [inline target/linux/ramips/image/mt7621.mk:258]: strange order, I'd have added metadata first and then checked size ...

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+&pcie0 {
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
+		ieee80211-freq-limit = <2400000 2500000>;
+
+		led {
+			led-active-low;
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:124]: What happens if this is specified without led-sources?

Comment (Sixtiz) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:124]: What does led-sources do and how do I know what value to put? LEDs seem to be working fine here...

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c (hunk 3 above)
Comment (Sixtiz) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c]: Used sge

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c (hunk 4 above)
Comment (Sixtiz) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c]: Updated

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c (hunk 6 above)
Comment (Sixtiz) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c]: Updated

File: target/linux/ramips/image/mt7621.mk (hunk 8 above)
Comment (Sixtiz) [reply target/linux/ramips/image/mt7621.mk]: Updated

File: target/linux/ramips/image/mt7621.mk (hunk 10 above)
Comment (Sixtiz) [reply target/linux/ramips/image/mt7621.mk]: Updated

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts (hunk 9 above)
Comment (Sixtiz) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts]: Updated

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts (hunk 7 above)
Comment (Sixtiz) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts]: Updated

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts (hunk 11 above)
Comment (Lucky1openwrt) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:124]: if you remove led-active-low is defaults to led-active-high & the LED's are inverted. 
on when meant to be off & off when meant to be on
never tried to remove it all with current driver but I'm guessing the led's won't work at all

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c (hunk 5 above)
Comment (Sixtiz) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c]: Updated

File: target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts (hunk 11 above)
Comment (Sixtiz) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:124]: I think he is referring to the keyword led-sources, not led-adcive-low. I can see in other dts files there is "led-sources = <2>" inserted...

Comment (adrianschmutzler) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:124]: If your implementation works correctly for both LEDs independent of each other, I'm perfectly fine. I just noted the missing led-sources lines, but I'm also not really sure what those do exactly either.
@mans0n Maybe you can help here?

Comment (mans0n) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:124]: IIRC led-sources and led-active-low correspond to gpio pin number and GPIO_ACTIVE_LOW, respectively. If led-sources is omitted, the default value 0 is used.

Comment (Sixtiz) [reply target/linux/ramips/dts/mt7621_dlink_dir-878-a1.dts:124]: > If your implementation works correctly for both LEDs independent of each other, I'm perfectly fine. I just noted the missing led-sources lines, but I'm also not really sure what those do exactly either.

Yes I confirm all LEDs seem to be working correctly and independently.

File: target/linux/ramips/image/mt7621.mk [hunk 12]
```diff
@@ -86,6 +86,11 @@ define Build/mitrastarimage
 	mv $@.new $@
 endef
 
+define Build/sge-uimage
+	uimage_padhdr -i $@ -o $@.new -l 96
+	mv $@.new $@
+endef
```
Comment (adrianschmutzler) [inline target/linux/ramips/image/mt7621.mk:92]: I wonder whether it's really necessary to define that again for every different padding required.
I'd consider to just merge that with mitrastarimage and fonfxcimage in image/Makefile.

I will create a patch for that separately ...

Comment (adrianschmutzler) [reply target/linux/ramips/image/mt7621.mk:92]: https://patchwork.ozlabs.org/project/openwrt/patch/20200710121623.1923-1-freifunk@adrianschmutzler.de/

With that, the sge-uimage recipe definition could be dropped and replaced by `uimage-padhdr 96` for the image build recipes.

**Comment by adrianschmutzler** [comment]
@ynezz The device part of the PR seems fine AFAICT. I don't want to delay this with the recipe patch I just brought up, so if you want to merge this as is I wouldn't mind to just cover the changes in my consolidation patch afterwards as well.

File: target/linux/ramips/image/mt7621.mk (hunk 12 above)
Comment (adrianschmutzler) [reply target/linux/ramips/image/mt7621.mk:92]: @Sixtiz I've pushed my patch now, so you can rebase onto it.

Comment (ynezz) [reply target/linux/ramips/image/mt7621.mk:92]: Please rebase and use that generic `uimage-padhdr 96` as suggested.

**Comment by ynezz** [comment]
Merged into my staging tree.

**Comment by mbc07** [comment]
Apart from minor differences, the following D-Link routers shares the same PCB, SoC, WiFi chips, RAM and flash memory:
* DIR-867 A1 (identical to DIR-878 A1 but limited to 3x3 streams by not soldering one of the antenna outputs of each WiFi chip)
* DIR-882 A1 (identical to DIR-878 A1 + 1 USB 2.0 port, 1 USB 3.0 port, 2 more LEDs)

Those have some more significant changes, but still shares the same PCB, SoC and WiFi chips:
* DIR-1760 A1 (identical to DIR-867 A1 + 1 USB 3.0 port, 1 more LED, 256 MB of RAM and 128 MB of NAND flash)
* DIR-1960 A1 (identical to DIR-1760 A1 but with 256QAM support disabled in firmware)
* DIR-2660 A1 (identical to DIR-882 A1 + 256 MB of RAM and 128 MB of NAND flash)

I also confirmed on a DIR-882 A1 that you can cross flash official firmwares from both DIR-867 and DIR-878 on it without any issues (as long as you use the emergency recovery mode) and the router still works normally, apart from missing USB support when using DIR-867/878 firmware. Haven't tried to cross-flash the DIR-1760/1960/2660 firmwares as those have NAND flash instead of SPI flash and double RAM so it very likely wouldn't work.

**TL;DR** what would be the next step to bring support to these? Looks like most of the work has already been done on this PR. Given the similarity between them, a shared DTS for DIR-867/878/882 and another shared DTS for DIR-1760/1960/2660 might be doable?

**Comment by Lucky1openwrt** [comment]
@mbc07 the Dir-878 & 1960 are both 4x4
or the models with NAND Flash you can start with this 
https://github.com/Lucky1openwrt/openwrt/commit/f9a25081a94dedd95b0a9e693c2f43d63d20b9e3
here is an early version for the DIR-1960
need's more testing tho

**Comment by mbc07** [comment]
> @mbc07 the Dir-878 & 1960 are both 4x4

Are you really sure about the DIR-1960 being 4x4? FCC Internal Photos of it shows one of the antenna outputs of each MT7615N chip not soldered just like the DIR-867, which I'm 100% sure is limited to 3x3 streams. D-Link website and FCC test reports also seem to supports this...

**Comment by Lucky1openwrt** [comment]
I'm Not 100% but here are some pics
DIR-1960 http://luckys.onmypc.net/openwrt/IMG_20200712_170138.jpg
DIR-1960 http://luckys.onmypc.net/openwrt/IMG_20200712_170049.jpg
DIR-878 http://luckys.onmypc.net/openwrt/IMG_20200712_170103.jpg

**Comment by mbc07** [comment]
All these models have 4 antennas, but you have to dig underneath the shields to see the differences. DIR-867/1760/1960 have a missing signal amplifier and each MT7615N has an unsoldered output: https://imgur.com/a/xOBJcwX

Meanwhile DIR-878/882/2660 have all 4 signal amplifiers and all outputs of both MT7615N chips wired to the 4 antennas: https://imgur.com/a/Tr0kKAd

DIR-878 3x3 streams and AC1900 limitation was just on the firmware, you could always crossflash it with DIR-882 firmware to fully utilize the hardware. Glad to see D-Link lifted this restriction on the most recent firmware...

**Comment by AlphaAnother** [comment]
Motorola's MR2600 and dir-882 use the same PCB, only one USB2.0 interface is missing.
This is the picture in the FCC report:
https://fccid.io/2AF5PMR2600/Internal-Photos/Internal-Photos-3706558

**Comment by Lucky1openwrt** [comment]
Ok I'm not going to strip it down that much I do know that the performance of the wifi is bad atm
maybe this has something to do with it
none of them connect to my EA8500 at a good speed but they do talk to each other & say 1733Mbut the performance seemed bad copying a file across it  but the 878 & 1960 are as good & bad as each other
